v3.0.0 advance midtrans config

Xaxxis opened the pull request:
Feature:
- new static method CoreApi, SnapApi, IrisApi, TransactionApi
- can change, TimeUnit for http connection settings
- can set global config via Midtrans class
- can add custom headers with config class
- validation for server-key empty, contain whitespace

Improvement:
- handle okhttp raw error/exception via MidtransError exception class

Notable changes:
- Migrate from retrofit to okhttp as http client

Commit:
handle new 3ds sample in Backend

Commit:
Update Java docs

Commit:
Update example: simple sample usage

Commit:
Update example: change initiate config, using builder

Commit:
Update example: change initiate config, using builder

Commit:
Update example: change snapapi using static method

Commit:
Update example: add new constant and DataMockup order-id prefix

Commit:
Increase version and change scope depedency

Commit:
Update config: add option can use Global config

Commit:
Update httpclient: no longer use retrofit, handle exception, options change timeunit. serverkey validation

Commit:
Change api class from retrofit using okhttp

Commit:
no longer use

Commit:
add message, status code, response body, and raw okhttp response

Commit:
Add new global config class

Commit:
add option builder method

Commit:
Change main api method from retrofit to new static method

Commit:
add utility method

Commit:
update functionality test

Commit:
Update how to use, and changes

Commit:
rename unit test class and update some method test logic

Commit:
Update java docs

Commit:
Fix typo

rizdaprasetya commented on README.md:
```diff
@@ -21,7 +21,7 @@ Maven:
 	  <groupId>com.midtrans</groupId>
 	  <artifactId>java-library</artifactId>
-	  <version>2.1.1</version>
+	  <version>3.1.1</version>
```
When working with `<major>.<minor>.<patch>` versioning format, if the major is increased, all the lower level should reset to 0. So this should be version `3.0.0`

rizdaprasetya commented on README.md:
```diff
@@ -59,8 +59,53 @@ Choose one that you think best for your unique needs.
 Get your client key and server key from [Midtrans Dashboard](https://dashboard.midtrans.com)
 
-Create API client object
+Create API client object, may you can see the [project's functional tests](library/src/test/java/com/midtrans/java) for more examples.
```
grammar, please fix to:
You can also check the ....

rizdaprasetya commented on README.md:
````diff
@@ -59,8 +59,53 @@ Choose one that you think best for your unique needs.
+JSONObject result = SnapApi.createTransaction(param, configOptions);
+```
+
+others alternative
````
maybe should be:
`#### Alternative way to initialize`

rizdaprasetya commented on README.md:
````diff
@@ -149,16 +204,34 @@ add new 4 params to use http proxy and network config,
+
 example: 
 ```java
 //TimeUnit for integer value is SECONDS
````
Since the example has changed to milisecond, this comment line no longer valid

rizdaprasetya commented on README.md:
````diff
@@ -181,13 +254,27 @@ coreApi.apiConfig().paymentOverrideNotification("https://example.com/test1,https
 ```
 When both `X-Append-Notification` and `X-Override-Notification` are used together then only override will be used.
 
+### Handling Error / Exception
+When using function that result in Midtrans API call e.g: CoreApi.chargeTransaction(...) or SnapApi.createTransaction(...) there's a chance it may throw error (MidtransError object), the error object will contains below properties that can be used as information to your error handling logic:
+```java
+try {
+    JSONObject result = CoreApi.chargeTransaction(param);
+} catch (MidtransError e) {
+    e.printStackTrace();
+    e.getStatusCode(); // basic error message string
+    e.getMessage(); // HTTP status code e.g: 400, 401, etc.
+    e.getResponseBody(); // API response body
+    e.getResponse(); // raw OkHttp response object
+}
+```
 #### CoreAPI Simple Sample Usage
 ```java
 
 import com.midtrans.Config;
 import com.midtrans.ConfigFactory;
-import com.midtrans.service.MidtransCoreApi;
+import com.midtrans.httpclient.CoreApi;
 import com.midtrans.httpclient.error.MidtransError;
+import com.midtrans.Midtrans;
````
Please check if import is still used, looks like Config & ConfigFactory is no longer used in the sample code? Maybe remove from import.

rizdaprasetya commented on CHANGELOG.md:
```diff
@@ -1,5 +1,19 @@
+- validation for server-key empty, contain whitespace
+
+Improvement:
+- handle okhttp raw error/exception via MidtransError exception class
+- No longer use retrofit, but still using okhttp as http client
```
Maybe separate:
> No longer use retrofit, but still using okhttp as http client

Into:
```
Notable changes:
- Migrate from retrofit to okhttp as http client
```

rizdaprasetya commented on example/src/main/java/com/midtrans/sample/SampleIrisApi.java:
```diff
@@ -0,0 +1,83 @@
+import static com.midtrans.sample.data.Constant.sandboxClientKey;
+import static com.midtrans.sample.data.Constant.sandboxServerKey;
+
+public class SampleIrisApi {
+
```
Is this expected, the class say IrisApp but the implementations of this file is mostly Snap?

Xaxxis replied:
Nope, it should be to iris sample implementation. Will change da

rizdaprasetya commented on example/src/main/java/com/midtrans/sample/controller/MobileSdkController.java:
```diff
@@ -1,30 +1,84 @@
+import static com.midtrans.sample.data.Constant.sandboxServerKey;
+
 @RestController
 public class MobileSdkController {
 
```
This is for mobile SDK backend (merchant server sample) right? Not for now, but for backlog:
- consider renaming the file/class into more descriptive, maybe:
> MobileSdkBackend

To avoid confusion merchant thinking that this implementation will automatically get them MobileSDK ready. Because they still need to setup Frontend mobile SDK client code on their app.

rizdaprasetya commented on example/src/main/java/com/midtrans/sample/controller/MobileSdkController.java:
```diff
@@ -1,30 +1,84 @@
+            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    @PostMapping(value = "/mobile-notification", produces = MediaType.APPLICATION_JSON_VALUE)
```
for backlog:
- consider renaming path to `/payment-status-notification-handler`

rizdaprasetya commented on library/src/main/java/com/midtrans/Config.java:
```diff
@@ -1,35 +1,39 @@
+@EqualsAndHashCode(callSuper = false)
 public class Config {
 
     private String SERVER_KEY;
     private String CLIENT_KEY;
     private String IRIS_MERCHANT_KEY;
```
I just realize this, is these always been all capitals since beginning? because the function name `setSERVER_KEY` is awkward and should be regular `setServerKey`.

In node JS capitalized var name is to indicate private static final vars, because NodeJS have no private static final. Not sure best practice in java. But in this case api keys are not private static final.

Well if it was always been capitals since beginning then we are forced to avoid change it, to avoid breaking changes. Or introduce/add new alias setter functions with the proper names.

rizdaprasetya commented on library/src/main/java/com/midtrans/ConfigFactory.java:
```diff
@@ -40,19 +40,17 @@ public MidtransCoreApiImpl getCoreApi() {
      * Get midtrans config.
      *
-     * @return {@link Config configuration} when connectng to midtrans API
+     * @return {@link Config configuration} when request to midtrans API
```
should be:
```
@return {@link Config configuration} that will be used by the HTTP Client when sending request to Midtrans API
````

rizdaprasetya commented on library/src/main/java/com/midtrans/utils/Utility.java:
```diff
@@ -0,0 +1,25 @@
+
+public final class Utility {
+
+    public static String base64(String serverKey) {
+        return "Basic " + Base64.getEncoder().encodeToString((serverKey + ":").getBytes(StandardCharsets.UTF_8));
+    }
```
minor, but function should be named `encodeBase64` for better readability.

rizdaprasetya commented on library/src/test/java/com/midtrans/java/CoreApiTest.java:
```diff
@@ -13,203 +13,189 @@
+        /* Using method with Global Config */
+        JSONObject result1 = CoreApi.cardToken(creditCard(cardNumberAccept, mainClientKey));
+        assert result1.has("token_id");
+        assert result1.getString("token_id").matches("[0-9]{6}-[0-9]{4}-[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}");
+        assert result1.getString("status_code").equals("200");
```
No need to check snap token format with regex. Just check if it is string

rizdaprasetya commented on library/src/test/java/com/midtrans/java/CoreApiTest.java:
```diff
@@ -13,203 +13,189 @@
+        try {
+            CoreApi.chargeTransaction(null);
+        } catch (MidtransError e) {
+            e.printStackTrace();
+            assert e.getMessage().contains("Midtrans API is returning API error. HTTP status code: 500 API response: {\"message\":\"An unexpected error occurred\"}");
```
Error message from API can change/improved, which if the API change it, this test will break. No need to assert the message value in this general 500 error case.

rizdaprasetya commented on library/src/test/java/com/midtrans/java/MidtransCoreApiTest.java:
```diff
@@ -0,0 +1,220 @@
+    public void cardToken() throws MidtransError {
+        JSONObject result = coreApi.cardToken(creditCard(cardNumberAccept));
+        assert result.has("token_id");
+        assert result.getString("token_id").matches("[0-9]{6}-[0-9]{4}-[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}");
```
no need to check token with regex, just check if it string

rizdaprasetya commented on library/src/test/java/com/midtrans/java/MidtransSnapApiTest.java:
```diff
@@ -0,0 +1,86 @@
+        try {
+            snapApi.createTransaction(dataMockup.miniDataMockUp());
+        } catch (MidtransError e) {
+            assert e.getMessage().equals("Midtrans API is returning API error. HTTP status code: 401 API response: {\"error_messages\":[\"Access denied due to unauthorized transaction, please check client or server key\",\"Visit https://snap-docs.midtrans.com/#request-headers for more details\"]}");
```
no need to check the exact whole API err message value, because it can change. Usually i'll check if it contains an important keyword only, in this case the message should contains `unauthorized`

rizdaprasetya commented on library/src/test/java/com/midtrans/java/mockupdata/Constant.java:
```diff
@@ -1,17 +1,47 @@
+    public static final int KEEP_ALIVE_DURATION = 400;
+
+    public static final String APPEND_NOTIFICATION = "https://example.com/";
+    public static final String OVERRIDE_NOTIFICATION = "https://google.com/";
```
Avoid using popular or big company website like `google.com` as notif url, because Google may block Midtrans server  if they think Midtrans is spamming their website with notif url.

`example.com` is fine

rizdaprasetya commented:
@Xaxxis overall nice improvements 👍 with some reviews above. Thanks!

Commit:
Update sample app demo

- Update IrisSampleClass
- Change setServerkey and clientKey method,
- add shopeepay and briva on ui demo,
- rename class mobile sdk controller,
- Fix Midtrans logo

Commit:
Fix from review: Change version

Commit:
Update and fix from review

- refactor and add new method for set credentialkey
- deprecate method setSERVER_KEY, setCLIENT_KEY, setIRIS_MERCHANT_KEY

Commit:
Update from review, change method name

Commit:
Update readme from review

rizdaprasetya reviewed (approved):
Looks good, all review solved. Go ahead merge & perform release 👍 
Thanks @Xaxxis !

Commit:
Update sample app, to use latest lib

Commit:
Update readme and pom.xml about repo migrate

Commit:
Merge branch 'master' into v3.1.1-Advance-Midtrans-Config
